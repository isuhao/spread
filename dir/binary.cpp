#include "binary.hpp"
#include <mangle/stream/servers/file_stream.hpp>
#include <mangle/stream/servers/outfile_stream.hpp>
#include <mangle/stream/servers/null_stream.hpp>
#include "hash/hash_stream.hpp"
#include <vector>

using namespace Spread;
using namespace std;
using namespace Mangle::Stream;

typedef Hash::DirMap DirMap;

const int MAGIC_NUMBER = 0x4d3bface;

static void fail(const std::string &msg) { throw std::runtime_error(msg); }
static void failDir(const std::string &msg)
{ fail("Error parsing directory file: " + msg); }

Hash Spread::Dir::read(Hash::DirMap &dir, StreamPtr strm)
{
  HashStream inf(strm);

  int tmp;

  // Read magic number
  inf.read(&tmp, 4);
  if(tmp != MAGIC_NUMBER)
    failDir("invalid magic number");

  // Number of hashes
  inf.read(&tmp, 4);

  // If someone throws us a malicious object, better to throw than
  // to crash.
  if(tmp > 1024*1024)
    failDir("too many directory hashes");

  Hash h;
  uint8_t hbuf[40];
  vector<char> buf;
  for(int i=0; i<tmp; i++)
    {
      // Read and parse hash
      inf.read(hbuf, 40);
      h.copy(hbuf);

      // String length + data
      int len = 0;
      inf.read(&len, 2);
      if(buf.size() < len)
        buf.resize(len);
      inf.read(&buf[0], len);

      // Store
      dir[string(&buf[0], len)] = h;
    }

  return inf.finish();
}

Hash Spread::Dir::write(const Hash::DirMap &dir, StreamPtr strm)
{
  HashStream ouf(strm);

  int tmp;

  tmp = MAGIC_NUMBER;
  ouf.write(&tmp, 4);

  tmp = dir.size();
  ouf.write(&tmp, 4);

  for(DirMap::const_iterator it = dir.begin();
      it != dir.end(); it++)
    {
      ouf.write(it->second.getData(), 40);

      const void *p = it->first.c_str();
      tmp = it->first.size();

      // Write size (2 bytes) first, then the string it self
      ouf.write(&tmp, 2);
      ouf.write(p, tmp);
    }

  return ouf.finish();
}

Hash Spread::Dir::read(Hash::DirMap &dir, const std::string &file)
{ return read(dir, FileStream::Open(file)); }

Hash Spread::Dir::write(const Hash::DirMap &dir, const std::string &file)
{ return write(dir, OutFileStream::Open(file)); }

Hash Spread::Dir::hash(const Hash::DirMap &dir)
{
  return write(dir, StreamPtr(new NullStream));
}
