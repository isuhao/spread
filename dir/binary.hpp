#ifndef __SPREAD_BINARY_DIR_HPP_
#define __SPREAD_BINARY_DIR_HPP_

#include <hash/hash.hpp>
#include <mangle/stream/stream.hpp>

namespace Spread
{
  namespace Dir
  {
    /* Read and write binary dir files. Returns the hash of the data
       read or written.
     */
    extern Hash read(Hash::DirMap &dir, const std::string &file);
    extern Hash write(const Hash::DirMap &dir, const std::string &file);

    // Stream versions
    extern Hash read(Hash::DirMap &dir, Mangle::Stream::StreamPtr strm);
    extern Hash write(const Hash::DirMap &dir, Mangle::Stream::StreamPtr strm);

    /* Return the hash of the dir, without actually writing the data
       anywhere.
     */
    extern Hash hash(const Hash::DirMap &dir);
  };
}
#endif
