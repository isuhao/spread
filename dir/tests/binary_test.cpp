#include "binary.hpp"
#include <iostream>

using namespace std;
using namespace Spread;

int main()
{
  Hash::DirMap dir;

  Hash h = Dir::read(dir, "test.bin");
  cout << "Read hash: " << h << endl;
  cout << "Elements: " << dir.size() << endl;
  cout << "Hash: " << Dir::hash(dir) << endl;
  cout << "name1: " << dir["name1"] << endl;
  cout << "dir/name2: " << dir["dir/name2"] << endl;

  return 0;
}
