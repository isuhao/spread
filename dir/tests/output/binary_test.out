Read hash: 795KeeO4JFV5WiZKYlYM-O4gAaUHG7IxwCvqqgJMwHRq
Elements: 2
Hash: 795KeeO4JFV5WiZKYlYM-O4gAaUHG7IxwCvqqgJMwHRq
name1: OqPmjeE85Dm6rKjRCr4xHyOB-hG6LspFzBK4nR0ppG2RdgY
dir/name2: zWyAFPZ5tztKheUOW7ifxqDG32EzGk9fK9kq8Ma1iAckAQ
